### PR TITLE
Colorize word-diff

### PR DESCRIPTION
**Pull request by kaste**
For the diff view, we have a 'word-diff' setting, but the output was annoying.

Let's fix that 😝 

mFpzzXJ9XW

Note that there is actually no one-size regex for word-diffs. Before, we just used `--word-diff` which uses a "default" which doesn't help in our context because it doesn't split enough. (E.g. it does not even split by word, and actually very often still by line.)  So hitting `w` now is a three state toggle, 'off' and then there come two regexes. The first one looks complicated and tries to match words, the second one (simply a `.`) matches single characters. 

Note that all three modes (t.i. the universal diff included) *can* look confusing. It is really to offer choices. 

Note that word diffs often suppress whitespace changes. The previous default did this, and it is probably okay. (That said I didn't find a regex that shows whitespace in a decent or clever way.)

Last: if we had a better ThemeGenerator abstraction this would we a really small PR.

### CODE REVIEW

**Comment by stoivo** [comment]
Look good, updated the popup like 

Screenshot 2019-10-22 at 22 30 07

File: core/commands/diff.py [hunk 1]
```diff
@@ -205,9 +265,26 @@ def _run(self):
         self.view.settings().set("git_savvy.diff_view.raw_diff", diff)
         text = prelude + '\n--\n' + diff
 
+        added_regions = []
+        removed_regions = []
+
+        if word_diff_regex:
+            def extractor(match):
+                a, b = match.span()
```
[the remark this answers is not in the document]
Comment (asfaltboy) [reply core/commands/diff.py]: Sounds more like C or Golang people... pythonista, myself included, prefer longer descriptive names... perhaps start and end would work better?

Ref: https://docs.python.org/3/library/re.html#re.Match.span

File: core/commands/diff.py [hunk 2]
```diff
@@ -223,7 +300,10 @@ def run(self, edit, setting):
 
         setting_str = "git_savvy.diff_view.{}".format(setting)
         current_mode = settings.get(setting_str)
-        next_mode = not current_mode
+        if setting == 'show_word_diff':
+            next_mode = (current_mode + 1) % len(WORD_DIFF_PATTERNS)
+        else:
+            next_mode = not current_mode
```
Comment (asfaltboy) [inline core/commands/diff.py]: I like this, but maybe it should be a separate command, e.g `GsDiffCycleSetting` ?

File: core/commands/diff.py [hunk 3]
```diff
@@ -205,9 +265,26 @@ def _run(self):
...
 
+        added_regions = []
+        removed_regions = []
+
+        if word_diff_regex:
+            def extractor(match):
+                a, b = match.span()
+                regions = added_regions if match.group()[1] == '+' else removed_regions
+                offset = len(added_regions) + len(removed_regions)
+                regions.append(sublime.Region(a - offset * 4, b - (offset + 1) * 4))
```
Comment (asfaltboy) [inline core/commands/diff.py]: Either I'm slow or it's late, or both.. but I can't figure out why we multiply by 4 ... any chance we could add a docstring here?

File: core/commands/diff.py [hunk 4]
```diff
@@ -205,9 +265,26 @@ def _run(self):
...
+                return match.group()[2:-2]
```
Comment (kaste) [inline core/commands/diff.py]: The `4` is here in `[2:-2]` in that we shorten the original text by `4` characters for each match. The original text has e.g. `{+foo+}` and we transform it to just `foo`.

**Comment by kaste** [comment]
Let's have a look at it again, folks!

**Comment by kaste** [comment]
LGTM. 

TODO: Refactor the ThemeGenerator thing, this basically adds 50+ lines noise to the file.

**Comment by kaste** [comment]
I just applied the fixup commit.

**Pull request merged**